lang/java/avro/src/main/java/org/apache/avro/Conversion.java

```diff
@@ -0,0 +1,171 @@
+ * <p>
+ * Instances of this class are added to GenericData to convert a logical type
+ * to a particular representation.
+ * <p>
+ * Implementations must provide:
+ * * {@link #getConvertedType()}: get the Java class used for the logical type
+ * * {@link #getLogicalTypeName()}: get the logical type this implements
+ * <p>
+ * Subclasses must also override all of the conversion methods for Avro's base
+ * types that are valid for the logical type, or else risk causing
+ * {@code UnsupportedOperationException} at runtime.
+ * <p>
+ * Optionally, use {@link #getRecommendedSchema()} to provide a Schema that
+ * will be used when a Schema is generated for the class returned by
+ * {@code getConvertedType}.
+ *
+ * @param <T> a Java type that generic data is converted to
+ */
+public abstract class Conversion<T> {
+
+  /**
+   * Return the Java class representing the logical type.
+   *
+   * @return a Java class returned by from methods and accepted by to methods
+   */
+  public abstract Class<T> getConvertedType();
+
+  /**
+   * Return the logical type this class converts.
+   *
+   * @return a String logical type name
+   */
+  public abstract String getLogicalTypeName();
+
+  public Schema getRecommendedSchema() {
+    throw new UnsupportedOperationException(
+        "No recommended schema for " + getLogicalTypeName());
+  }
+
+  public T fromBoolean(Boolean value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromBoolean is not supported for " + type.getName());
+  }
+
+  public T fromInt(Integer value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromInt is not supported for " + type.getName());
+  }
+
+  public T fromLong(Long value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromLong is not supported for " + type.getName());
+  }
+
+  public T fromFloat(Float value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromFloat is not supported for " + type.getName());
+  }
+
+  public T fromDouble(Double value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromDouble is not supported for " + type.getName());
+  }
+
+  public T fromCharSequence(CharSequence value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromCharSequence is not supported for " + type.getName());
+  }
+
+  public T fromEnumSymbol(GenericEnumSymbol value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromEnumSymbol is not supported for " + type.getName());
+  }
+
+  public T fromFixed(GenericFixed value, Schema schema, LogicalType type) {
+    throw new UnsupportedOperationException(
+        "fromFixed is not supported for " + type.getName());
+  }
+
+  public T fromBytes(ByteBuffer value, Schema schema, LogicalType type)  {
```

Comment:
I added tests for decimal as fixed and bytes in reflect and the runtime type was actually ByteBuffer rather than bytes. I think that makes sense because without a "java-class" property, the default should be the generic type. If there is a "java-class" then the [logical type is invalid|https://github.com/apache/avro/pull/29/files#diff-c035d2730b60f23ca63a93db0e8dcfcbR71].